File: 07_studying/python/graphs/src/graphs/graph.py
from collections import defaultdict
from typing import TypeVar, Protocol, Generic, Iterable

A = TypeVar("A")
B = TypeVar("B")


class Visitor(Protocol, Generic[A, B]):
    def __call__(self, node: int, context: B | None, graph: "Graph") -> A:
        pass


class Graph(Protocol):
    @classmethod
    def from_edges(cls, edges: list[(int, int)]) -> "Graph":
        pass

    def get_nodes(self) -> set[int]:
        pass

    def get_neighbors(self, node: int) -> set[int]:
        pass

    def dfs_preorder(self: "Graph", start: int, visitor: Visitor[A, B]) -> Iterable[A]:
        visited = set()
        stack = [(start, None)]
        while stack:
            node = stack.pop()
            if node not in visited:
                visited.add(node)
                yield visitor(node, self)
                for neighbor in self.get_neighbors(node):
                    stack.append(neighbor)

    def dfs_postorder(self: "Graph", start: int, visitor: Visitor[A]) -> Iterable[A]:
        visited = set()
        stack = [start]
        while stack:
            node, neighbors = stack.pop()
            if node not in visited:
                visited.add(node)
                yield visitor(node, neighbors)
                for neighbor in self.get_neighbors(start):
                    stack.append((neighbor, self.get_neighbors(neighbor)))

File: 07_studying/rust/graphs/src/graph/graph.rs
use rstest::Context;

use super::Node;
use super::edge::Edge;
use std::collections::HashSet;
use std::process::Output;

trait Graph<const DIRECTED: bool, E: Edge<DIRECTED>> {
    fn from_edges<const N: usize>(edges: [E; N]) -> Self;
    fn get_nodes(&self) -> HashSet<Node>;
    fn get_edges(&self) -> HashSet<E>;
}

trait DirectedGraph<E: Edge<true>>: Graph<true, E> {
    fn get_parents(&self, node: Node) -> HashSet<Node>;
    fn get_children(&self, node: Node) -> HashSet<Node>;
}

trait UndirectedGraph<E: Edge<false>>: Graph<false, E> {
    fn get_neighbors(&self, node: Node) -> HashSet<Node>;
}

enum SearchType {
    DFS
    BFS,
}

trait Search<const DIRECTED: bool, G: Graph<DIRECTED, E>, Context, Output, E: Edge<DIRECTED>> {
    fn visit(node: Node, graph: G, context: Context) -> Output;
    fn context_update(node: Node, parent: Node, graph: G, context: Context) -> Context;
}

// fn dfs(graph: G, start: Node, initial_context: Context) -> Vec<Output> {
//     let mut visited: HashSet<Node> = HashSet::new();
//     let mut stack: Vec<(Node, Context)> = Vec::new();
//     let mut output: Vec<Output> = Vec::new();
//     stack.push((start, initial_context));
//     while let Some((node, context)) = stack.pop() {
//         if visited.contains(&node) {
//             continue;
//         }
//         visited.insert(node);
//         output.push(Self::visit(node, graph, context));
//         if DIRECTED {
//         } else {
//             let neighbors = graph.get_neighbors(node);
//             for neighbor in neighbors {
//                 let new_context = Self::context_update(neighbor, node, graph, context);
//                 stack.push((neighbor, new_context));
//             }
//         }
//     }
//     output
// }
